Comorphisms/CASL2Prenex.hs

```diff
@@ -0,0 +1,150 @@
+
+import Common.AS_Annotation
+import Common.ProofTree
+
+data CASL2Prenex = CASL2Prenex deriving Show
+
+instance Language CASL2Prenex where
+    language_name CASL2Prenex = "CASL2Prenex"
+
+instance Comorphism CASL2Prenex
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree
+               CASL CASL_Sublogics
+               CASLBasicSpec CASLFORMULA SYMB_ITEMS SYMB_MAP_ITEMS
+               CASLSign
+               CASLMor
+               Symbol RawSymbol ProofTree where
+    sourceLogic CASL2Prenex = CASL
+    sourceSublogic CASL2Prenex = SL.caslTop --TODO
+    targetLogic CASL2Prenex = CASL
+    mapSublogic CASL2Prenex sl = Just sl -- TODO
+    map_theory CASL2Prenex = mapTheory
+    map_morphism CASL2Prenex = error "nyi"
+    map_sentence CASL2Prenex _ _ = error "nyi"
+    map_symbol CASL2Prenex _ = Set.singleton . id
+    has_model_expansion CASL2Prenex = True -- check
+    is_weakly_amalgamable CASL2Prenex = True --check
+
+-- remove equivalences and implications,
+-- turn arbitrary junctions into binary ones
+preprocessSen :: CASLFORMULA -> CASLFORMULA
+preprocessSen sen = case sen of
+ Relation sen1 Equivalence sen2 _ -> let -- remove equivalences
+   sen1' = preprocessSen sen1
+   sen2' = preprocessSen sen2
+  in Junction Con [Junction Dis [Negation sen1' nullRange, sen2'] nullRange,
+                   Junction Dis [Negation sen2' nullRange, sen1'] nullRange
+                  ] nullRange
+ Relation sen1 Implication sen2 _ -> let -- remove implications
+   sen1' = preprocessSen sen1
+   sen2' = preprocessSen sen2
+  in Junction Dis [Negation sen1' nullRange, sen2'] nullRange
+ Relation sen1 RevImpl sen2 _ -> let -- remove reverse implications
+   sen1' = preprocessSen sen1
+   sen2' = preprocessSen sen2
+  in Junction Dis [Negation sen2' nullRange, sen1'] nullRange
+ Junction j sens _ -> let -- only binary junctions
+   sens' = map preprocessSen sens
+   bindLeft [] = Junction j [] nullRange
+   bindLeft [x] = Junction j [x] nullRange
+   bindLeft [x,y] = Junction j [x,y] nullRange
+   bindLeft (x:y:s') = Junction j [x, bindLeft (y:s')] nullRange
+  in bindLeft sens'
+ Quantification q vdecls qsen _ -> let -- recursion
+   qsen' = preprocessSen qsen
+  in Quantification q vdecls qsen' nullRange
+ _ -> sen
+
+-- function 2: computePNF
+-- see page 292 in LPL.pdf
+-- but instead of checking that a variable occurs free in a formula
+-- I do a alpha renaming such that this property is ensured
+
+computePNF :: Int -> CASLFORMULA -> (Int, CASLFORMULA)
+computePNF n sen = case sen of
+ Negation nsen _ -> let
+   (n', nsen') = computePNF n nsen
+   (n'', nsen'') = case nsen' of
+    Quantification q vdecls qsen _ -> -- deMorgan
+       (n', Quantification (dualQuant q) vdecls
+              (Negation qsen nullRange) nullRange)
+    _ -> (n', Negation nsen' nullRange)
+  in (n'', nsen'')
```

Comment:
This does not work. Quantifiers may remain inside a negation. E.g.
```
spec Prenex =
  sort s
  preds p,q:s
  . not (exists x:s . ( forall y:s . p(x) \/ q(y)))
end
```
You need to go recursively through `nsen'` and dualise all its quantifiers.